core/config/solver_config.hpp

```diff
@@ -17,19 +19,20 @@ namespace config {
 
 
 template <typename SolverParam>
-inline void common_solver_parse(SolverParam& params, const pnode& config,
-                                const registry& context,
-                                type_descriptor td_for_child)
+void common_solver_parse(SolverParam& params, const pnode& config,
+                         const registry& context, type_descriptor td_for_child,
+                         std::set<std::string>& allowed_keys)
 {
-    if (auto& obj = config.get("generated_preconditioner")) {
+    if (auto& obj =
```

Comment:
We could simplify the diff and improve readability by wrapping the get_config_node call in a lambda that inserts the config and allowed_keys parameters. This goes for almost every change in this PR.

Reply:
The difference is most visible for long names like `symbolic_factorization`, where the change forces the if condition into two lines, while previously we only had a single line.

I do like the config_structure idea, I would probably call it a view or a decorator from the class design perspective, but essentially they do the same thing.

Destructors avoiding throwing exceptions only matters if the exception is thrown when unrolling another exception, and that is not a place to be calling Gingko code anyways.

Reply:
two lines to me is not a big issue to prevent them.
Yes, I mean three of them are the same thing, which just give different way to pass allowed_keys.
Although I use macro a lot, I would also like to avoid the macro usage if there is an alternative one without introducing too many stacks.